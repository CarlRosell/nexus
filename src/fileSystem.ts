import path from "path";

export class FileSystem {
  protected static cachedInstance: FileSystem;

  static getInstance() {
    if (!this.cachedInstance) {
      this.cachedInstance = new FileSystem();
    }
    return this.cachedInstance;
  }

  protected get util() {
    return require("util") as typeof import("util");
  }

  protected get fs() {
    return require("fs") as typeof import("fs");
  }

  protected get mkdir() {
    return this.util.promisify(this.fs.mkdir);
  }

  protected get readFile() {
    return this.util.promisify(this.fs.readFile);
  }

  protected get writeFile() {
    return this.util.promisify(this.fs.writeFile);
  }

  protected get unlink() {
    return this.util.promisify(this.fs.unlink);
  }

  protected ensureAbsolute(filePath: string) {
    if (!path.isAbsolute(filePath)) {
      throw new Error(
        `GraphQL Nexus: Expected an absolute path, saw ${filePath}`
      );
    }
    return filePath;
  }

  getFile(fileName: string) {
    return this.readFile(this.ensureAbsolute(fileName), "utf8");
  }

  /**
   * Unlinks & writes a new file - useful when we are generating the
   * type definitions, as this causes the ts server to know the file has changed.
   * @param path
   * @param toSave
   */
  async replaceFile(filePath: string, toSave: string) {
    if (this.fs.existsSync(this.ensureAbsolute(filePath))) {
      await this.unlink(filePath);
    }
    const dirName = path.dirname(filePath);
    if (!this.fs.existsSync(dirName)) {
      await this.mkdir(dirName, { recursive: true });
    }
    await this.writeFile(filePath, toSave);
  }
}
