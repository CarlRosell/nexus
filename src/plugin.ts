import { SchemaConfig } from "./builder";
import {
  GraphQLFieldResolver,
  GraphQLResolveInfo,
  GraphQLFieldConfig,
} from "graphql";
import { withNexusSymbol, NexusTypes, Omit } from "./definitions/_types";
import { isPromise } from "./utils";
import { NexusObjectTypeConfig } from "./definitions/objectType";
import { NexusOutputFieldConfig } from "./definitions/definitionBlocks";

export class PluginDef {
  constructor(readonly config: PluginConfig) {}
}
withNexusSymbol(PluginDef, NexusTypes.Plugin);

export type PluginVisitAfter = (
  result: any,
  root: any,
  args: any,
  ctx: any,
  info: GraphQLResolveInfo,
  breakVal: typeof BREAK_RESULT_VAL
) => any;

export type PluginVisitBefore = (
  root: any,
  args: any,
  ctx: any,
  info: GraphQLResolveInfo,
  nextVal: typeof NEXT_RESULT_VAL
) => any;

export type PluginVisitor = {
  after?: PluginVisitAfter;
  before?: PluginVisitBefore;
};

export type PluginDefinitionInfo = {
  /**
   * The name of the type we're applying the plugin to
   */
  typeName: string;
  /**
   * The name of the field we're applying the plugin to
   */
  fieldName: string;
  /**
   * The root-level SchemaConfig passed
   */
  nexusSchemaConfig: Omit<SchemaConfig, "types"> & Record<string, unknown>;
  /**
   * The config provided to the Nexus type containing the field.
   * Will not exist if this is a non-Nexus GraphQL type.
   */
  nexusTypeConfig?: Omit<NexusObjectTypeConfig<string>, "definition"> &
    Record<string, unknown>;
  /**
   * The config provided to the Nexus type containing the field.
   * Will not exist if this is a non-Nexus GraphQL type.
   */
  nexusFieldConfig?: Omit<NexusOutputFieldConfig<string, string>, "resolve"> &
    Record<string, unknown>;
  /**
   * Info about the GraphQL Field we're decorating.
   * Always guaranteed to exist, even for non-Nexus GraphQL types
   */
  graphqlFieldConfig: Omit<GraphQLFieldConfig<any, any>, "resolve">;
  /**
   * If we need to collect/reference metadata during the
   * plugin middleware definition stage, you can use this object.
   *
   * An example use would be to collect all fields that have a "validation"
   * property for their input to reference in runtime. After the schema is complete,
   * this object is frozen so it is not abused at runtime.
   */
  mutableObj: Record<string, any>;
};

export interface RootTypingImport {
  /**
   * File path to import the type from.
   */
  path: string;
  /**
   * Name of the type we want to reference in the `path`
   */
  name: string;
  /**
   * Name we want the imported type to be referenced as
   */
  alias?: string;
}

export interface PluginConfig {
  /**
   * A name for the plugin, useful for errors, etc.
   */
  name: string;
  /**
   * A description for the plugin
   */
  description?: string;
  /**
   * Any type definitions we want to add to the schema
   */
  schemaTypes?: string;
  /**
   * Any type definitions we want to add to the type definition option
   */
  typeDefTypes?: string;
  /**
   * Any type definitions we want to add to the field definitions
   */
  fieldDefTypes?: string;
  /**
   * Any extensions to the GraphQLInfoObject (do we need this?)
   */
  // infoExtension?: string;
  /**
   * Any types which should exist as standalone declarations to support this type
   */
  localTypes?: string;
  /**
   * Definition for the plugin. This will be executed against every
   * output type field on the schema.
   */
  pluginDefinition(pluginInfo: PluginDefinitionInfo): PluginVisitor | void;
}

/**
 * A plugin defines configuration which can document additional metadata options
 * for a type definition.
 *
 * Ultimately everything comes down to the "resolve" function
 * which executes the fields, and our plugin system will take that into account.
 *
 * You can specify options which can be defined on the schema,
 * the type or the plugin. The config from each of these will be
 * passed in during schema construction time, and used to augment the field as necessary.
 *
 * You can either return a function, with the new defintion of a resolver implementation,
 * or you can return an "enter" / "leave" pairing which will wrap the pre-execution of the
 * resolver and the "result" of the resolver, respectively.
 */
export function plugin(config: PluginConfig) {
  return new PluginDef(config);
}

/**
 * On "before" plugins, the nextFn allows us to skip to the next resolver
 * in the chain
 */
const NEXT_RESULT_VAL = Object.freeze({});

/**
 * On "after" plugins, the breakerFn allows us to early-return, skipping the
 * rest of the plugin stack. Useful when encountering errors
 */
const BREAK_RESULT_VAL = Object.freeze({});

/**
 * Wraps a resolver with a plugin, "before" the resolver executes.
 * Returns a new resolver, which may subsequently be wrapped
 */
export function wrapPluginsBefore(
  resolver: GraphQLFieldResolver<any, any>,
  beforePlugins: [string, PluginVisitBefore][]
): GraphQLFieldResolver<any, any> {
  return async (root, args, ctx, info) => {
    for (let i = 0; i < beforePlugins.length; i++) {
      const [name, before] = beforePlugins[i];
      let result = before(root, args, ctx, info, NEXT_RESULT_VAL);
      if (isPromise(result)) {
        result = await result;
      }
      if (result === NEXT_RESULT_VAL) {
        continue;
      }
      if (result === undefined) {
        throw new Error(
          `Nexus: Expected return value from plugin ${name}:before, saw undefined`
        );
      }
      return result;
    }
    return resolver(root, args, ctx, info);
  };
}

/**
 * Wraps a resolver with a plugin, "after" the resolver executes.
 * May return a new value for the return type
 */
export function wrapPluginsAfter(
  resolver: GraphQLFieldResolver<any, any>,
  afterPlugins: [string, PluginVisitAfter][]
): GraphQLFieldResolver<any, any> {
  return async (root, args, ctx, info) => {
    let finalResult: any;
    try {
      finalResult = resolver(root, args, ctx, info);
      if (isPromise(finalResult)) {
        finalResult = await finalResult;
      }
    } catch (e) {
      finalResult = e;
    }
    for (let i = 0; i < afterPlugins.length; i++) {
      const [name, after] = afterPlugins[i];
      let returnVal = after(
        finalResult,
        root,
        args,
        ctx,
        info,
        BREAK_RESULT_VAL
      );
      if (isPromise(returnVal)) {
        returnVal = await returnVal;
      }
      if (returnVal === BREAK_RESULT_VAL) {
        return finalResult;
      }
      if (returnVal === undefined) {
        throw new Error(
          `Nexus: Expected return value from plugin ${name}:after, saw undefined`
        );
      }
    }
    return finalResult;
  };
}
