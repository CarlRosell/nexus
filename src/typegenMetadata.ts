import path from "path";
import { Options } from "prettier";
import { GraphQLSchema, lexicographicSortSchema, printSchema } from "graphql";
import { TypegenPrinter } from "./typegen";
import { SDL_HEADER, TYPEGEN_HEADER } from "./lang";
import { typegenAutoConfig } from "./typegenAutoConfig";
import {
  BuilderConfig,
  TypegenInfo,
  NexusSchema,
  SchemaBuilder,
} from "./builder";
import { FileSystem } from "./fileSystem";

export type TypegenFormatFn = (
  content: string,
  type: "types" | "schema"
) => string | Promise<string>;

/**
 * Passed into the SchemaBuilder, this keeps track of any necessary
 * field / type metadata we need to be aware of when building the
 * generated types and/or SDL artifact, including but not limited to:
 */
export class TypegenMetadata {
  protected config: BuilderConfig;
  protected typegenFile: string = "";
  protected sortedSchema: GraphQLSchema;
  protected finalPrettierConfig?: Options;

  constructor(
    protected builder: SchemaBuilder,
    protected nexusSchema: NexusSchema
  ) {
    this.sortedSchema = lexicographicSortSchema(nexusSchema);
    const config = (this.config = builder.getConfig());
    if (config.outputs !== false && config.shouldGenerateArtifacts !== false) {
      if (config.outputs.typegen) {
        this.typegenFile = this.assertAbsolutePath(
          config.outputs.typegen,
          "outputs.typegen"
        );
      }
    }
  }

  getNexusSchema() {
    return this.nexusSchema;
  }

  getBuilder() {
    return this.builder;
  }

  getTypegenFile() {
    return this.typegenFile;
  }

  getSortedSchema() {
    return this.sortedSchema;
  }

  /**
   * Generates the type definitions
   */
  protected async generateTypesFile(): Promise<string> {
    const typegenInfo = await this.getTypegenInfo();
    return new TypegenPrinter(this, typegenInfo).print();
  }

  /**
   * Generates the artifacts of the build based on what we
   * know about the schema and how it was defined.
   */
  async generateArtifacts() {
    if (this.config.outputs) {
      if (this.config.outputs.schema) {
        await this.writeTypeFile(
          "schema",
          this.generateSchemaFile(this.sortedSchema),
          this.config.outputs.schema
        );
      }
      if (this.typegenFile) {
        const value = await this.generateTypesFile();
        await this.writeTypeFile("types", value, this.typegenFile);
      }
    }
  }

  protected async writeTypeFile(
    type: "schema" | "types",
    output: string,
    filePath: string
  ) {
    const [toSave, existing] = await Promise.all([
      this.formatTypegen(output, type),
      FileSystem.getInstance()
        .getFile(filePath)
        .catch(() => ""),
    ]);
    if (toSave !== existing) {
      return FileSystem.getInstance().replaceFile(filePath, toSave);
    }
  }

  /**
   * Generates the schema, adding any directives as necessary
   */
  protected generateSchemaFile(schema: GraphQLSchema): string {
    let printedSchema = printSchema(schema);
    return [SDL_HEADER, printedSchema].join("\n\n");
  }

  async getTypegenInfo(): Promise<TypegenInfo> {
    if (this.config.typegenConfig) {
      if (this.config.typegenAutoConfig) {
        console.warn(
          `Only one of typegenConfig and typegenAutoConfig should be specified, ignoring typegenConfig`
        );
      }
      return this.config.typegenConfig(this.sortedSchema, this.typegenFile);
    }
    if (this.config.typegenAutoConfig) {
      return typegenAutoConfig(this.config.typegenAutoConfig)(
        this.sortedSchema,
        this.typegenFile
      );
    }
    return {
      headers: [TYPEGEN_HEADER],
      imports: [],
      contextType: "any",
      backingTypeMap: {},
    };
  }

  /**
   * Takes "content", which is a string representing the GraphQL schema
   * or the TypeScript types (as indicated by the "fileType"), and formats them
   * with prettier if it is installed.
   * @param content
   * @param fileType
   */
  protected async formatTypegen(content: string, fileType: "types" | "schema") {
    if (this.config.formatTypegen instanceof Function) {
      return this.config.formatTypegen(content, fileType);
    }
    if (!this.config.prettierConfig) {
      return content;
    }
    try {
      const prettierConfig = this.config.prettierConfig;
      const prettier = require("prettier") as typeof import("prettier");
      if (typeof prettierConfig === "string") {
        if (!this.finalPrettierConfig) {
          this.finalPrettierConfig = JSON.parse(
            await FileSystem.getInstance().getFile(prettierConfig)
          ) as Options;
        }
      } else {
        this.finalPrettierConfig = prettierConfig;
      }
      return prettier.format(content, {
        ...this.finalPrettierConfig,
        parser: fileType === "schema" ? "graphql" : "typescript",
      });
    } catch (e) {
      console.error(e);
    }
    return content;
  }

  protected assertAbsolutePath(pathName: string, property: string) {
    if (!path.isAbsolute(pathName)) {
      throw new Error(
        `Expected path for ${property} to be a string, saw ${pathName}`
      );
    }
    return pathName;
  }
}
