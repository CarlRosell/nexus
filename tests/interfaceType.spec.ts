import { graphql } from "graphql";
import path from "path";
import {
  deferred,
  interfaceType,
  makeSchema,
  objectType,
  queryField,
} from "../src/core";
import { FileSystem } from "../src/fileSystem";

describe("interfaceType", () => {
  let spy: jest.SpyInstance;
  beforeEach(() => {
    jest.clearAllMocks();
    spy = jest
      .spyOn(FileSystem.prototype, "replaceFile")
      .mockImplementation(async () => null);
  });

  it("can be implemented by object types", async () => {
    const dfd = deferred();
    const schema = makeSchema({
      types: [
        interfaceType({
          name: "Node",
          definition(t) {
            t.id("id");
            t.resolveType(() => null);
          },
        }),
        objectType({
          name: "User",
          definition(t) {
            t.implements("Node");
            t.string("name");
          },
        }),
        queryField("user", {
          type: "User",
          resolve: () => ({ id: `User:1`, name: "Test User" }),
        }),
      ],
      onReady: dfd.resolve,
      outputs: {
        schema: path.join(__dirname, "interfaceTypeTest.graphql"),
        typegen: false,
      },
      shouldGenerateArtifacts: true,
    });
    await dfd.promise;
    expect(spy).toBeCalledTimes(1);
    expect(spy.mock.calls[0]).toMatchSnapshot();
    expect(
      await graphql(
        schema,
        `
          {
            user {
              id
              name
            }
          }
        `
      )
    ).toMatchSnapshot();
  });
});
