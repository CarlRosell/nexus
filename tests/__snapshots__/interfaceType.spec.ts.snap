// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interfaceType can be implemented by object types 1`] = `
Array [
  "/Users/tgriesser/Github/oss/nexus/nexus/tests/interfaceTypeTest.graphql",
  "### This file was autogenerated by GraphQL Nexus
### Do not make changes to this file directly


interface Node {
  id: ID!
}

type Query {
  user: User!
}

type User implements Node {
  id: ID!
  name: String!
}
",
]
`;

exports[`interfaceType can be implemented by object types 2`] = `
Object {
  "data": Object {
    "user": Object {
      "id": "User:1",
      "name": "Test User",
    },
  },
}
`;
